OpenRA.Game/Network/Nat.cs

```diff
@@ -0,0 +1,101 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Threading;
+using Mono.Nat;
+
+namespace OpenRA.Network
+{
+	public enum NatStatus { Enabled, Disabled, NotSupported }
+
+	public class Nat
+	{
+		public static NatStatus Status => NatUtility.IsSearching ? natDevice != null ? NatStatus.Enabled : NatStatus.NotSupported : NatStatus.Disabled;
+
+		static Mapping mapping;
+		static INatDevice natDevice;
+		static bool initialized;
+
+		public static void Initialize()
+		{
+			if (initialized)
+				return;
+
+			if (Game.Settings.Server.DiscoverNatDevices)
+			{
+				NatUtility.DeviceFound += DeviceFound;
+				NatUtility.StartDiscovery();
+			}
+
+			initialized = true;
+		}
+
+		static readonly SemaphoreSlim Locker = new SemaphoreSlim(1, 1);
+
+		static async void DeviceFound(object sender, DeviceEventArgs args)
+		{
+			await Locker.WaitAsync();
+			try
+			{
+				// Only interact with one at a time. Some support both UPnP and NAT-PMP.
+				natDevice = args.Device;
+
+				Log.Write("nat", "Device found: {0}", natDevice.NatProtocol);
```

Comment:
I believe this should be `natDevice.DeviceEndpoint`, and then `NatProtocol` in the Type line below? `NatProtocol` and `GetType().Name` are reporting basically the same thing, neither of which actually identify the device!

Reply:
Updated.